Update the driver.search API endpoint

jnormington opened the pull request:
It appears that at sometime M$ have changed the API for the drive search which means that it is currently broken right now within the Geckoboard spreadsheet workflow

However with the break after reviewing the documentation it appears they have introduced a new endpoint which can search all files that a user has access to. This is exactly what we want, as it means that users will no longer need to move their files around to the root directory of their drive just to satisfy Geckoboard spreadsheets integration.

https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_search?view=odsp-graph-online

So instead of updating the search to use the new /me/drive/root/search(q='{search-query}')

I've instead updated to use to the new search all files user has access /me/drive/search(q='{search-query}')

This endpoint does have pagination, although I'm not sure we need to implement, firstly a page can contain almost upto 200 resources in a single response, also returning anymore than a single page on the panel would increase the wait time for the user to see list of files, but we have and do push the user already to refine the search to get the specific file they are looking for (with gsheets for example), which is also why i think as is should be fine.

jnormington committed:
Update the driver.search API endpoint

It appears that at sometime M$ have changed the API for the drive
search which means that it is currently broken right now.

However with the break after reviewing the documentation it appears
they have introduced a new endpoint which can search all files
that a user has access to. This is exactly what we want, as it means
that users will no longer need to move their files around just to
satify Geckoboard spreadsheets integration.

https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_search?view=odsp-graph-online

So instead of updating the search to use the new
/me/drive/root/search(q='{search-query}')

I've instead updated to use to the new search all files user has access
/me/drive/search(q='{search-query}')

ferglor commented on drive.go:
```diff
@@ -144,15 +145,15 @@ func (ds *DriveService) ListChildren(ctx context.Context, driveID string, filter
-		q = q + "&filter=" + url.QueryEscape(filter)
+// Search the hierarchy of items for all items the user has access to.
+func (ds *DriveService) Search(ctx context.Context, q string) (*Items, *http.Response, error) {
+	// We need to escape single quotes with a double single
```
I guess drop the `a` from this? E.g. `We need to escape single quotes with double single quotes, otherwise the request will fail with an unterminated error` ?

glenjamin commented on drive.go:
```diff
@@ -144,15 +145,15 @@ func (ds *DriveService) ListChildren(ctx context.Context, driveID string, filter
+	// We need to escape single quotes with a double single
+	// quotes, otherwise API will fail with unterminated
+	if strings.Contains(q, "'") {
+		q = strings.ReplaceAll(q, "'", "''")
```
perhaps this could be `pathEscapeIncludingQuotes` and have it encode single quotes as well as calling `url.PathEscape` in one helper?

EDIT: does the url-encoding of single quotes work? I could imagine that being slightly less weird to see in practice

jnormington replied:
great will do 👍

jnormington replied:
@glenjamin if I understand the edit question the url encoding isn't just for the single quotes - but other special characters that would break if the user tried them and were not encoded.

glenjamin commented on drive.go:
```diff
@@ -144,15 +145,15 @@ func (ds *DriveService) ListChildren(ctx context.Context, driveID string, filter
 	}
 
-	req, err := ds.newRequest("GET", "/drive/root/view.search?q="+q, nil, nil)
+	req, err := ds.newRequest("GET", fmt.Sprintf("/drive/search(q='%s')", url.PathEscape(q)), nil, nil)
```
what a weird API URL 🙄

jnormington replied:
> I think this looks okay!
> 
> Do you also need to vendor your dependencies? `go mod vendor`, and commit those?

@ferglor Ah I'm not sure to be honest, should we still be committing the vendor directory with go modules ?

glenjamin replied:
I think we only need to vendor for applications, not libraries

ferglor reviewed (approved):
I think this looks okay!

Do you also need to vendor your dependencies? `go mod vendor`, and commit those?

jnormington committed:
Extract the quoting and escaping of the query into helper func

glenjamin commented on drive.go:
```diff
@@ -165,3 +160,13 @@ func (ds *DriveService) Search(ctx context.Context, q string, filter string) (*I
+func (ds *DriveService) pathEscapeIncludingQuotes(query string) string {
+	// We need to escape single quotes with double single
+	// quotes, otherwise API will fail with unterminated
+	if strings.Contains(query, "'") {
```
I don't think the `if` is needed, you can just call `ReplaceAll` and it'll do that check internally I'd expect

glenjamin reviewed (approved):
Is there a reference for the query encoding rules anywhere? if so it would be good to include a link

jnormington committed:
Sanitize the query to remove some "dangerous" characters

This isn't documented anywhere but the API returns a 400
response that there are dangerous characters in the query, and so
fails this won't be a nice experience for the customers trying to
search for files.

So through testing we've identified a small list of characters which
are danagerous and would fail. This means its possible to make an
empty search criteria - which doesn't break anything and would
return the first page of results with the xlsx files filtered out.

This also instead of trying to escape the single quotes it now also
treats that as a dangerous character, as it seems that removing it
returns the files with quotes inbetween file names, so removing it
returns the relevant files still.

glenjamin commented on drive.go:
```diff
@@ -5,6 +5,11 @@ import (
+)
+
+var (
+	dangerousQueryCharsRegexp = regexp.MustCompile("\\*|<|>|'|\\?|\\\\|\\/")
```
using a raw string might make this easier to read:
```go
`\*|<|>|'|\?|\|/`
```

glenjamin commented on drive_test.go:
```diff
@@ -243,3 +246,134 @@ func TestListDriveChildrenInvalid(t *testing.T) {
+		}
+
+		for _, spec := range specs {
+			mux.HandleFunc("/drive/search(q='"+spec.out+"')", fileWrapperHandler("fixtures/drive.search.valid.json", 200))
```
🏄 I'd be tempted to make this handler be at `/drive`, and have it record the `r.URL.Path` it saw, and then the assertions could be `expected: path, got: other-path`, instead of a failing case giving a 404.

glenjamin commented on drive_test.go:
```diff
@@ -243,3 +246,134 @@ func TestListDriveChildrenInvalid(t *testing.T) {
+		}
+	})
+
+	t.Run("escapes when search contains single quote", func(t *testing.T) {
```
i think this case is invalid now?
